Modules/gcmodule.c

```diff
@@ -1191,80 +1251,260 @@ handle_resurrected_objects(PyGC_Head *unreachable, PyGC_Head* still_unreachable,
- * collection process works. */
-static Py_ssize_t
-gc_collect_main(PyThreadState *tstate, int generation,
-                Py_ssize_t *n_collected, Py_ssize_t *n_uncollectable,
-                int nofail)
+#define UNTRACK_TUPLES 1
+#define UNTRACK_DICTS 2
+
+static void
+gc_collect_region(PyThreadState *tstate,
+                  PyGC_Head *from,
+                  PyGC_Head *to,
+                  int untrack,
+                  struct gc_collection_stats *stats);
+
+static inline Py_ssize_t
+gc_list_set_space(PyGC_Head *list, uintptr_t space)
 {
-    GC_STAT_ADD(generation, collections, 1);
+    Py_ssize_t size = 0;
+    PyGC_Head *gc;
+    for (gc = GC_NEXT(list); gc != list; gc = GC_NEXT(gc)) {
+        gc_set_old_space(gc, space);
+        size++;
+    }
+    return size;
+}
+
+static void
+add_stats(GCState *gcstate, int gen, struct gc_collection_stats *stats)
+{
+    gcstate->generation_stats[gen].collected += stats->collected;
+    gcstate->generation_stats[gen].uncollectable += stats->uncollectable;
+    gcstate->generation_stats[gen].collections += 1;
+}
+
+static void
+gc_collect_young(PyThreadState *tstate,
+                 struct gc_collection_stats *stats)
+{
+    GCState *gcstate = &tstate->interp->gc;
+    PyGC_Head *young = &gcstate->young.head;
+    PyGC_Head *aging = &gcstate->old[gcstate->aging_space].head;
 #ifdef Py_STATS
-    if (_Py_stats) {
-        _Py_stats->object_stats.object_visits = 0;
+    {
+        Py_ssize_t count = 0;
+        PyGC_Head *gc;
+        for (gc = GC_NEXT(young); gc != young; gc = GC_NEXT(gc)) {
+            count++;
+        }
+        GC_STAT_ADD(0, objects_queued, count);
     }
 #endif
-    int i;
-    Py_ssize_t m = 0; /* # objects collected */
-    Py_ssize_t n = 0; /* # unreachable objects that couldn't be collected */
-    PyGC_Head *young; /* the generation we are examining */
-    PyGC_Head *old; /* next older generation */
-    PyGC_Head unreachable; /* non-problematic unreachable trash */
-    PyGC_Head finalizers;  /* objects with, & reachable from, __del__ */
-    PyGC_Head *gc;
-    _PyTime_t t1 = 0;   /* initialize to prevent a compiler warning */
-    GCState *gcstate = &tstate->interp->gc;
+    PyGC_Head survivors;
+    gc_list_init(&survivors);
+    gc_collect_region(tstate, young, &survivors, UNTRACK_TUPLES, stats);
+    Py_ssize_t surivivor_count = 0;
+    if (gcstate->aging_space) {
+        /* objects in aging space have bit set, so we set it here */
+        surivivor_count = gc_list_set_space(&survivors, 1);
+    }
+    else {
+        PyGC_Head *gc;
+        for (gc = GC_NEXT(&survivors); gc != &survivors; gc = GC_NEXT(gc)) {
+#ifdef GC_DEBUG
+            assert(gc_old_space(gc) == 0);
+#endif
+            surivivor_count++;
+        }
+    }
+    gc_list_merge(&survivors, aging);
+    validate_old(gcstate);
+    gcstate->young.count = 0;
+    gcstate->old[gcstate->aging_space].count++;
+    gcstate->incremental_gc_progress -= surivivor_count;
```

Comment:
I wonder if this can be fine tuned some more beyond just accounting for the `surivivor_count` here, maybe more aggressively decrementing `incremental_gc_progress` here and then aggressively increasing it when the `survivor_count` is high in the incremental collection.  Kind of with an idea of if collection is productive we should do more of it, but if it's not we should do less of it.  Or something else entirely :)

Here's kind of a pathological program which isn't doing great, it takes 140,864,398 allocations before the garbage link list is cleared:

```
import gc
import weakref

gc.freeze()
print('----------------')
class LinkedList:
    def __init__(self, next = None, prev = None):
        self.next = next
        if next is not None:
            next.prev = self
        self.prev = prev
        if prev is not None:
            prev.next = self

head = LinkedList()
wr = weakref.ref(head)
for i in range(100000):
    if i % 1000 == 0: print('##############', i)
    head = LinkedList(head, head.prev)

head2 = LinkedList()
wr2 = weakref.ref(head)
for i in range(100000):
    if i % 1000 == 0: print('##############', i)
    head2 = LinkedList(head2, head2.prev)

print(wr)
del head
del head2

count = 0
while True:
    head = LinkedList()
    tail = LinkedList(head, head)
    count += 2
    if wr() is None and wr2() is None:
        print('dead', count)
        break
```

Reply:
Also FWIW current 3.13 seems to need 69,545,980 allocations to free the garbage (3.8 which is my system Python seemed to need 34,441,446)